### ISSUE TITLE
 --registrar:legacystatic not working

### ISSUE THREAD

**Issue by alexmantaut** [xamarin/xamarin-macios#4721]
I am trying to compile a project that used a third party legacy library for Xamarin (it is not possible to modify the library)

After updating to the latest version of VS, XCode and Xamarin iOS, the project stopped compiling with an MT4116 error.

Doing some research, it appears that the default registrar has been changed in new versions of Xamarin.

Following this [documentation]( https://docs.microsoft.com/en-us/xamarin/ios/internals/registrar), I tried to revert to the old registrars to compile the project, but got the error text in the report.

My question would be: have legacyregistrars been deprecated? If so, how can I compile this code, without modifying the library? Can I downgrade to an old version of Xamarin iOS and that will make the registrars work? Is there a better alternative? 

### Steps to Reproduce

1. Added --registrar:legacystatic option to iOS build, Additional mtouch arguments
2. Tried to compile

### Expected Behavior
I expected the code to compile

### Actual Behavior
Error MT0020: The valid options for '--registrar' are 'static, dynamic or default'. (MT0020)
### Environment

```
=== Visual Studio Community 2017 for Mac ===

Version 7.6.1 (build 9)
Installation UUID: 1d5cc319-43ac-43a7-84a2-11ca6a842e98
Runtime:
	Mono 5.12.0.301 (2018-02/4fe3280bba1) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 4.4.1.178 (master / eeaeb7e6)

	Package version: 512000301

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Version: 2.0.5
SDK: /usr/local/share/dotnet/sdk/2.1.4/Sdks
SDK Version: 2.1.4
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.12.0/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

'/Applications/Xamarin Profiler.app' not found

=== Xamarin.Android ===

Version: 8.2.0.16 (Visual Studio Community)
Android SDK: /Users/alexmantaut/Library/Android/sdk
	Supported Android versions:
		5.0 (API level 21)
		7.0 (API level 24)

SDK Tools Version: 26.0.1
SDK Platform Tools Version: 25.0.5
SDK Build Tools Version: 24.0.0

Java SDK: /usr
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Apple Developer Tools ===

Xcode 9.4.1 (14161)
Build 9F2000

=== Xamarin.Mac ===

Xamarin.Mac not installed. Can't find /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/Version.

=== Xamarin.iOS ===

Version: 11.14.0.13 (Visual Studio Community)
Hash: 373c313a
Branch: HEAD
Build date: 2018-07-24 23:47:12-0400

=== Build Information ===

Release ID: 706010009
Git revision: 1916291c332d942d45f3d8a952cc315f7864abef
Build date: 2018-08-22 19:19:11+00
Build branch: release-7.6
Xamarin extensions: 53cd52a227bb91315389d4860f821fa4c3b396b3

=== Operating System ===

Mac OS X 10.13.6
Darwin 17.7.0 Darwin Kernel Version 17.7.0
    Thu Jun 21 22:53:14 PDT 2018
    root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
```

**Comment by rolfbjarne** [comment]
The legacy registrars have been gone for over a year (and even before they only worked in very specific scenarios).

It's difficult to say how to fix this without seeing the actual problem you're facing: would it be possible to get a [verbose build log](https://docs.microsoft.com/en-us/xamarin/cross-platform/troubleshooting/questions/version-logs#verbose-build-output-logs) (with the MT4116 error)?

**Comment by alexmantaut** [comment]
Thanks for the answer, I tried to build the project with verbose build, I am attaching the logs.

The error I get is:

MTOUCH : error MT4116: Could not register the assembly 'ShinobiCharts': error MT4118: Cannot register two managed types ('ShinobiCharts.SChartSeriesCrosshairTooltipDrawerProtocol, ShinobiCharts' and 'SChartSeriesCrosshairTooltipDrawerProtocol, ShinobiCharts') with the same native name ('SChartSeriesCrosshairTooltipDrawerProtocol').

[build.log](https://github.com/xamarin/xamarin-macios/files/2331220/build.log)

**Comment by rolfbjarne** [comment]
`ShinobiCharts` is a commonly used third-party library, and should be compatible with Xamarin.iOS. Is there any particular reason you can't update to their latest version?

**Comment by alexmantaut** [comment]
The main reason I cannot update is that the project is using an old version of the library... I tried the new version and some interfaces are different, which would mean rewriting a lot of code to no benefit for my client...

I am wondering if downgrading xamarin ios would allow me to compile code as it used to do before... In which version were the registrars changed?

**Comment by rolfbjarne** [comment]
No, downgrading won't work, because only the Classic API supported the legacy registrars, and your app is not using the Classic API (it's using the Unified API - which is the right thing to do).

The problem with the Classic API is that it did not support 64-bit code, which means that any Classic app won't run on iOS 11+ (since Apple dropped 32-bit support in iOS 11).

However, it _might_ be possible to rewrite the ShinobiCharts.dll so that it works (depending on exactly what's wrong with it). Can you share it with us so that I can have a look?

**Comment by rolfbjarne** [comment]
It seems to be possible to fix the library.

I've created a tool that modifies the library: https://github.com/rolfbjarne/fix-binding-library

To fix the library:

* Clone https://github.com/rolfbjarne/fix-binding-library locally
* Checkout the `shinobicharts-v2.8.1` branch (https://github.com/rolfbjarne/fix-binding-library/tree/shinobicharts-v2.8.1).
* Run `make` to build the library.
* Execute `mono bin/Debug/fix-binding-library.exe /path/to/ShinobiCharts.dll`

The library will be modified in place, and the build should now complete successfully.

**Comment by rolfbjarne** [comment]
Closing since this is a known breaking change (#4065), and a solution has been provided.
